share/man/man1/malheur.1
.\" Automatically generated by Pod::Man 2.27 (Pod::Simple 3.28)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "MALHEUR 1"
.TH MALHEUR 1 "2014-11-18" "malheur 0.5.4" "User Manual"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
malheur \- automatic analysis of malware behavior
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBmalheur\fR [\fB\-hrvV\fR] [\fB\-m\fR \fImaldir\fR] [\fB\-o\fR \fIoutfile\fR] [\fB\-d\fR \fIdumpfile\fR] \fIaction\fR \fIdataset\fR
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBmalheur\fR is a tool for the automatic analysis of malware behavior (program
behavior recorded from malicious software in a sandbox environment).  The
tool has been designed to support the regular analysis of malicious software
and the development of detection and defense measures.  \fBmalheur\fR allows
for identifying novel classes of malware with similar behavior and assigning
unknown malware to discovered classes.  It supports four basic actions for
analysis which can be applied to reports of recorded behavior:
.IP "\fIExtraction of prototypes\fR." 4
.IX Item "Extraction of prototypes."
From a given set of reports, \fBmalheur\fR identifies a subset of prototypes
representative for the full data set.  The prototypes provide a quick
overview of recorded behavior and can be used to guide manual inspection.
.IP "\fIClustering of behavior\fR." 4
.IX Item "Clustering of behavior."
\&\fBmalheur\fR automatically identifies groups (clusters) of reports containing
similar behavior.  Clustering allows for discovering novel classes of
malware and provides the basis for crafting specific detection and defense
mechanisms, such as anti-virus signatures.
.IP "\fIClassification of behavior\fR." 4
.IX Item "Classification of behavior."
Based on a set of previously clustered reports, \fBmalheur\fR is able to assign
unknown behavior to known groups of malware.  Classification enables
identifying novel and unknown variants of malware and can be used to filter
program behavior prior to manual inspection.
.IP "\fIIncremental analysis\fR." 4
.IX Item "Incremental analysis."
\&\fBmalheur\fR can be applied incrementally for analysis of large data
sets. By processing reports in chunks, the run-time as well as memory
requirements can be significantly reduced. This renders long-term
application of \fBmalheur\fR feasible, for example for daily analysis 
of incoming malware programs.
.PP
A detailed description of these techniques as well as technical 
background on analysis of malicious software is provided in the 
following articles:
.IP "\(bu" 4
\&\fIAutomatic Analysis of Malware Behavior using Machine Learning\fR.
Konrad Rieck, Philipp Trinius, Carsten Willems, and Thorsten Holz.
Journal of Computer Security (\s-1JCS\s0), 19(4) 639\-668, 2011.
.IP "\(bu" 4
\&\fIA Malware Instruction Set for Behavior-Based Analysis\fR.
Philipp Trinius, Carsten Willems, Thorsten Holz, and Konrad Rieck.
Technical report \s-1TR\-2009\-07,\s0 University of Mannheim, 2009
.PP
The input of \fBmalheur\fR is a \fIdataset\fR containing reports of malware
behavior. The \fIdataset\fR is provided either as a directory or a
compressed archive containing the reports. \fBmalheur\fR supports the
following formats for compressed archives: tar.gz, zip, pax and cpio.
A \fIreport\fR is a textual document describing the recorded activity of
a malware program, where individual events are separated by delimiter
characters, such as white space or carriage return.  The events in a
report are expected to be in sequential order if n\-grams with n\ >\ 1 are extracted. If the behavior is represented using the
malware instruction set (\s-1MIST\s0) further options may be selected.
.PP
The result of an analysis is written to \fIoutfile\fR, a textual file
containing columns corresponding to particular analysis results.  By
default \fIoutfile\fR is set to \fImalheur.out\fR.
.PP
The configuration and internal state of \fBmalheur\fR are stored in the
directory \fImaldir\fR. If this directory does not exist, it is created
and the system wide configuration is copied. A detailed description of
the malheur configuration is provided in \*(L"\s-1CONFIGURATION\*(R"\s0. By
default \fImaldir\fR is set to \fI~/.malheur\fR.
.SH "ACTIONS & OPTIONS"
.IX Header "ACTIONS & OPTIONS"
\&\fBmalheur\fR supported different \fIactions\fR for analysis of a
\&\fIdataset\fR.  For all actions the reports are first mapped to a
high-dimensional vector space, such that each report is represented as
a feature vector. Recorded events or n\-grams of events are reflected
in different dimensions and the dissimilarity of behavior can be
assessed geometrically by computing distances and angles.
.IP "\fBdistance\fR" 12
.IX Item "distance"
If this action is specified, \fBmalheur\fR computes a distance matrix for
\&\fIdataset\fR. The entry (i,j) of this matrix reflects the distance
(dissimilarity) of the reports i and j. The distance values lie in the
range 0 to \fIsqrt\fR\|(2). The distance matrix is written to \fIoutfile\fR.
.IP "\fBprototype\fR" 12
.IX Item "prototype"
If this action is specified, \fBmalheur\fR determines a set of
\&\fIprototypes\fR representing \fIdataset\fR. The prototypes are selected
from the contained reports, such that the distance from any report to
its nearest prototype is minimized. The prototype assignment of
\&\fIdataset\fR is written to \fIoutfile\fR.
.IP "\fBcluster\fR" 12
.IX Item "cluster"
If this action is specified, \fBmalheur\fR performs a \fIclustering\fR of
\&\fIdataset\fR. The clustering is first determined on prototypes and then
propagated to all reports. Small clusters with too few members are
merged in a \fIrejection cluster\fR. The prototypes representing accepted
clusters are stored as internal state of \fBmalheur\fR for later
classification and incremental analysis. The clustering of \fIdataset\fR
is written to \fIoutfile\fR.
.IP "\fBclassify\fR" 12
.IX Item "classify"
If this action is specified, \fBmalheur\fR performs a \fIclassification\fR
of \fIdataset\fR. Each report is either assigned to the nearest prototype
of a known cluster or rejected as unknown. This action requires that a
clustering has been performed beforehand and an internal state of
\&\fBmalheur\fR exists. The classification of \fIdataset\fR is written to
\&\fIoutfile\fR.
.IP "\fBincrement\fR" 12
.IX Item "increment"
If this action is specified, \fBmalheur\fR performs an \fIincremental
analysis\fR of the reports. The reports are first classified to known
clusters as in the action \fIclassify\fR. Reports rejected from
classification are then clustered as in the action \fIcluster\fR. The
prototypes of the accepted clusters and the rejected reports are
written to the internal state of \fBmalheur\fR for further incremental
analysis. The classification and clustering of \fIdataset\fR are written
to \fIoutfile\fR.
.IP "\fBprotodist\fR" 12
.IX Item "protodist"
If this action is specified, \fBmalheur\fR computes a distance matrix for
prototypes.  The entry (i,j) of this matrix reflects the distance
(dissimilarity) of the prototypes i and j.  The distance values lie in the
range 0 to \fIsqrt\fR\|(2).  This action requires that either a prototype
extraction, a clustering or an incremental analysis has been performed
beforehand.  The distance matrix is written to \fIoutfile\fR.
.PP
\&\fBmalheur\fR also supports the following command-line \fIoptions\fR which
are used to further control the analysis process
.IP "\fB\-m\fR \fImaldir\fR" 14
.IX Item "-m maldir"
This option specifies the malheur directory \fImaldir\fR which holds the
configuration and internal state of \fBmalheur\fR. If the directory does
not exist, it is created and the system wide configuration is copied.
.IP "\fB\-o\fR \fIoutfile\fR" 14
.IX Item "-o outfile"
This option specifies the output file \fIoutfile\fR for analysis. The
file is created during analysis and the results are stored in textual
form.
.IP "\fB\-d\fR \fIdumpfile\fR" 14
.IX Item "-d dumpfile"
If this option is set, \fBmalheur\fR dumps the extracted feature vectors to
\&\fIdumpfile\fR in libsvm format.  Note that the output is not fully compatible
with all libsvm loaders and might need some manual refinement.
.IP "\fB\-r\fR" 14
.IX Item "-r"
This option resets the internal state of \fBmalheur\fR. Prototypes of
clusters and rejected reports from previous runs of \fBmalheur\fR are
removed.
.IP "\fB\-v\fR" 14
.IX Item "-v"
This option is used to increase the verbosity of \fBmalheur\fR during
analysis, where the verbosity level corresponds to the number 
of \*(L"\-v\*(R" options.
.IP "\fB\-h\fR" 14
.IX Item "-h"
This option prints a brief help screen.
.IP "\fB\-V\fR" 14
.IX Item "-V"
This option prints a version and copyright string.
.SH "CONFIGURATION"
.IX Header "CONFIGURATION"
The configuration file \fImalheur.cfg\fR in \fImaldir\fR determines how the
malware behavior in a data set is analysis. The configuration contains
five groups of settings which are described in the following.
.PP
All parameters of the configuration can be also specified on the
command line.  That is, if a parameter is defined in the configuration as
\&\fBxx = \*(L"yy\*(R";\fR in the group \fBzz\fR, it can be alternatively supplied as a

malheur/src/malheur.c
/*
 * MALHEUR - Automatic Analysis of Malware Behavior
 * Copyright (c) 2009-2012 Konrad Rieck (konrad@mlsec.org)
 * University of Goettingen, Berlin Institute of Technology
 * --
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 3 of the License, or (at your
 * option) any later version.  This program is distributed without any
 * warranty. See the GNU General Public License for more details. 
 * --
 */

#include "config.h"
#include "malheur.h"
#include "common.h"
#include "mconfig.h"
#include "ftable.h"
#include "fmath.h"
#include "proto.h"
#include "util.h"
#include "cluster.h"
#include "class.h"
#include "export.h"

/* Global variables */
int verbose = 0;
config_t cfg;

/* Global variables */
static char *output_file = OUTPUT_FILE;
static char malheur_dir[MAX_PATH_LEN];
static char **input_files = NULL;
static int input_len = 0;
static int reset = FALSE;
static int save = TRUE;
static char *fvec_dump = NULL;
static char *pvec_dump = NULL;
static malheur_action_t action = PROTOTYPE;
static malheur_cfg_t mcfg;
static malheur_state_t mstate;

/* Option string */
#define OPTSTRING       "nro:m:hvVd:p:"

/**
 * Array of options of getopt_long()
 */
static struct option longopts[] = {
   { "maldir",          1, NULL, 'm' },
   { "outfile",         1, NULL, 'o' },
   { "reset",           0, NULL, 'r' },
   { "nostate",         0, NULL, 'n' },
   { "verbose",         0, NULL, 'v' }, 
   { "version",         0, NULL, 'V' },
   { "help",            0, NULL, 'h' },
   { "dump",            1, NULL, 'd' },
   { "dumproto",		1, NULL, 'p' },
   /* start of config options */   
   { "input.format",           1, NULL, 1001 },
   { "input.mist_level",       1, NULL, 1002 },   
   { "input.mist_rlen",        1, NULL, 1003 },   
   { "input.mist_tlen",        1, NULL, 1004 },   
   { "features.ngram_delim",   1, NULL, 1005 },   
   { "features.ngram_len",     1, NULL, 1006 },
   { "features.vect_embed",    1, NULL, 1007 },
   { "features.lookup_table",  1, NULL, 1008 },
   { "prototypes.max_dist",    1, NULL, 1009 },   
   { "prototypes.max_num",     1, NULL, 1010 },   
   { "classify.max_dist",      1, NULL, 1011 },   
   { "cluster.link_mode",      1, NULL, 1012 },   
   { "cluster.min_dist",       1, NULL, 1013 },   
   { "cluster.reject_num",     1, NULL, 1014 },
   { "cluster.shared_ngrams",  1, NULL, 1015 },
   /* end of config options */
   { NULL,              0, NULL, 0 }
};

/**
 * Print usage of command line tool
 * @param argc Number of arguments
 * @param argv Argument values
 */
static void print_usage(void)
{
    printf("Usage: malheur [options] <action> <dataset>\n"
        "Actions:\n"
        "  distance       Compute distance matrix for malware reports\n"
        "  prototype      Extract prototypes from malware reports\n"
        "  protodist      Compute distance matrix for prototypes\n"        
        "  cluster        Cluster malware reports into similar groups\n"
        "  classify       Classify malware reports using labeled prototypes\n"
        "  increment      Incremental analysis of malware reports\n"
        "  info           Print information about internal state of Malheur\n"
        "Options:\n"
        "  -m <maldir>    Set malheur directory. [%s]\n"
        "  -o <outfile>   Set output file for analysis. [%s]\n"
        "  -d <dumpfile>  Dump feature vectors to file in libsvm format.\n"
        "  -r             Reset internal state of Malheur.\n"
        "  -n             Don't save internal state of Malher.\n"
        "  -v             Increase verbosity.\n"
        "  -V             Print version and copyright.\n"
        "  -h             Print this help screen.\n"
        "See manual page malheur(1) for more information.\n",
       malheur_dir, output_file);
}

/**
 * Parse command line options
 * @param argc Number of arguments
 * @param argv Argument values
 */
static void parse_options(int argc, char **argv)
{
    int ch;
    
    /* reset getopt */
    optind = 0;
    
    while ((ch = getopt_long(argc, argv, OPTSTRING, longopts, NULL)) != -1) {
        switch (ch) {
        case 'n': 
            save = FALSE;
            break;
        case 'r':
            reset = TRUE;
            break;
        case 'v':
        case 'm':
            /* Empty. See load_config() */
            break;
        case 'o':
            output_file = optarg;
            break;
        case 'V':
            malheur_version(stdout);
            exit(EXIT_SUCCESS);
            break;
        case 'd':
            fvec_dump = optarg;
            break;
		case 'p':
            pvec_dump = optarg;
			break;
        case 'h':
        case '?':
            print_usage();
            exit(EXIT_SUCCESS);
            break;

        /* long options */
        case 1001:
            config_set_string(&cfg, "input.format", optarg);    
            break;
        case 1002:
            config_set_int(&cfg, "input.mist_level", atoi(optarg));    
            break;
        case 1003:
            config_set_int(&cfg, "input.mist_rlen", atoi(optarg));    
            break;
        case 1004:
            config_set_int(&cfg, "input.mist_tlen", atoi(optarg));    
            break;
        case 1005:
            config_set_string(&cfg, "features.ngram_delim", optarg);    
            break;
        case 1006:
            config_set_int(&cfg, "features.ngram_len", atoi(optarg));    
            break;
        case 1007:
            config_set_string(&cfg, "features.vect_embed", optarg);    
            break;
        case 1008:
            config_set_int(&cfg, "features.lookup_table", atoi(optarg));    
            break;
        case 1009:
            config_set_float(&cfg, "prototypes.max_dist", atof(optarg));    
            break;
        case 1010:
            config_set_int(&cfg, "prototypes.max_num", atoi(optarg));    
            break;
        case 1011:
            config_set_float(&cfg, "classify.max_dist", atof(optarg));    
            break;
        case 1012:
            config_set_string(&cfg, "cluster.link_mode", optarg);    
            break;
        case 1013:
            config_set_float(&cfg, "cluster.min_dist", atof(optarg));    
            break;
        case 1014:
            config_set_int(&cfg, "cluster.reject_num", atoi(optarg));    
            break;
        case 1015:
            config_set_int(&cfg, "cluster.shared_ngrams", atoi(optarg));    
            break;
        }
    }

    /* Check configuration */
    config_check(&cfg);

    argc -= optind;
    argv += optind;

    if (argc < 1)
        fatal("the <action> argument is required");

    /* Argument: action */
    if (!strcasecmp(argv[0], "prototype")) {
        action = PROTOTYPE;
    } else if (!strcasecmp(argv[0], "distance")) {
        action = DISTANCE;
    } else if (!strcasecmp(argv[0], "cluster")) {
        action = CLUSTER;
    } else if (!strcasecmp(argv[0], "classify")) {
        action = CLASSIFY;
    } else if (!strcasecmp(argv[0], "increment")) {
        action = INCREMENT;
    } else if (!strcasecmp(argv[0], "protodist")) {
        action = PROTODIST;
    } else if (!strcasecmp(argv[0], "info")) {
        action = INFO;
    } else {
        fatal("Unknown analysis action '%s'", argv[0]);
    }
    
    if (argc < 2 && action != PROTODIST && action != INFO) 
        fatal("the <dataset> argument is required");

    /* Assign input files */
    input_files = argv + 1;
    input_len = argc - 1;
}

/**
 * Load configuration
 * @param argc Number of arguments
 * @param argv Argument values
 */
static void load_config(int argc, char **argv)
{
    char cfg_path[MAX_PATH_LEN];
    int ch;

    /* Prepare dir */
    snprintf(malheur_dir, MAX_PATH_LEN, "%s", MALHEUR_DIR);

    /* Check for config file in command line */
    while ((ch = getopt_long(argc, argv, OPTSTRING, longopts, NULL)) != -1) {
        switch (ch) {
            case 'm':
                strncpy(malheur_dir, optarg, MAX_PATH_LEN);
                malheur_dir[MAX_PATH_LEN - 1] = 0;
                break;
            case 'v':
                verbose++;
                break;
            case '?':
            default:
                /* empty */
                break;
        }
    }

    /* Prepare malheur files */
    snprintf(cfg_path, MAX_PATH_LEN, "%s/%s", malheur_dir, CONFIG_FILE);

    /* Check for directories and files */
    if (access(malheur_dir, F_OK))
        fatal("Malheur directory '%s'", malheur_dir);

    /* Copy configuration file */
    if (access(cfg_path, R_OK)) {
        if (verbose > 0)
            printf("Copying configuration to '%s'.\n", cfg_path);
        copy_file(GLOBAL_CONFIG_FILE, cfg_path);
    }

    /* Init and load configuration */
    config_init(&cfg);
    if (config_read_file(&cfg, cfg_path) != CONFIG_TRUE)
        fatal("Could not read configuration (%s in line %d)",
              config_error_text(&cfg), config_error_line(&cfg));

    /* Check configuration */
    config_check(&cfg);
    if (verbose > 1)
        config_print(&cfg);
}


/**
 * Initialize malheur tool
 * @param argc Number of arguments
 * @param argv Argument values
 */
static void malheur_init(int argc, char **argv)
{
    int lookup;
    double shared;

    /* Load configuration */
    load_config(argc, argv);

    /* Parse options */
    parse_options(argc, argv);

    /* Prepare malheur files */
    snprintf(mcfg.reject_file, MAX_PATH_LEN, "%s/%s", malheur_dir, REJECT_FILE);
    snprintf(mcfg.config_file, MAX_PATH_LEN,"%s/%s", malheur_dir, CONFIG_FILE);
    snprintf(mcfg.proto_file, MAX_PATH_LEN, "%s/%s", malheur_dir, PROTO_FILE);
    snprintf(mcfg.state_file, MAX_PATH_LEN, "%s/%s", malheur_dir, STATE_FILE);

    /* Init feature lookup table */
    config_lookup_int(&cfg, "features.lookup_table", &lookup);
    config_lookup_float(&cfg, "cluster.shared_ngrams", &shared);
    if (lookup || shared > 0.0) {
        ftable_init();
    }
    
    /* Reset current state */
    if (reset) {
        unlink(mcfg.reject_file);
        unlink(mcfg.proto_file);
        unlink(mcfg.state_file);
    }

    memset(&mstate, 0, sizeof(mstate));
}


/**
 * Loads data from archives/directories into feature vectors 
 * @return array of feature vectors
 */
static farray_t *malheur_load()
{
    farray_t *fa = NULL;
    for (int i = 0; i < input_len; i++) {
        /* Argument: Input */
        if (access(input_files[i], R_OK)) {
            warning("Could not access '%s'", input_files[i]);
            continue;
        }

        farray_t *f = farray_extract(input_files[i]);
        fa = farray_merge(fa, f);
    }
    
    if (!fa) 
        fatal("No data available.");
    
    /* Dump feature vectors to file */
    if (fvec_dump)
        farray_save_libsvm_file(fa, fvec_dump);

	if (pvec_dump) {
		assign_t *as;
		farray_t *pr;

		/* Extract prototypes */
		pr = proto_extract(fa, &as);
		farray_save_libsvm_file(pr, pvec_dump);
	}

    return fa;
}

/**
 * Saves the internal Malheur state. The state is used during incremental
 * analysis to distinguig clusters obtained during different runs
 * @param run Current run of analysis
 * @param proto Number of prototypes
 * @param rej Number of rejected reports
 */
static void malheur_save_state()
{
    FILE *f;

    if (verbose > 0)
        printf("Saving internal state to '%s'.\n", mcfg.state_file);
    
    f = fopen(mcfg.state_file, "w");
    if (!f) {
        error("Could not open state file '%s'.", mcfg.state_file);
        return;
    }
        
    fprintf(f, "run = %u\nprototypes = %u\nrejected = %u\n", 
            mstate.run, mstate.num_proto, mstate.num_reject);
    
    fclose(f);
}

/**
 * Loads the internal Malheur state. The state is used during incremental
 * analysis to distinguig clusters obtained during different runs
 */
static int malheur_load_state()
{
    FILE *f;
    int ret;
    
    if (access(mcfg.state_file, R_OK))
        return FALSE;
    
    f = fopen(mcfg.state_file, "r");
    if (!f) {
        error("Could not open state file '%s'.", mcfg.state_file);
        return FALSE;
    }
        
    ret = fscanf(f, "run = %u\nprototypes = %u\nrejected = %u\n", 
                &mstate.run, &mstate.num_proto, &mstate.num_reject);
    
    if (ret != 3) {
        error("Could not parse state file '%s'.", mcfg.state_file);
        return FALSE;
    }
    
    fclose(f);
    return TRUE;
}

/**
 * Determines prototypes for the given malware reports
 */
static void malheur_prototype()
{
    assign_t *as;
    farray_t *fa, *pr;
    
    /* Load data */
    fa = malheur_load();

    /* Extract prototypes */
    pr = proto_extract(fa, &as);
    if (verbose > 1)
        farray_print(pr);

    /* Save prototypes */
    if (save)
        farray_save_file(pr, mcfg.proto_file);

    /* Export prototypes */
    export_proto(pr, fa, as, output_file);

    /* Clean up */
    assign_destroy(as);
    farray_destroy(pr);
    farray_destroy(fa);
}

/**
 * Clusters the given malware reports
 */
static void malheur_cluster()
{
    assign_t *as;
    farray_t *fa, *pr, *pn, *re;

    /* Load data */
    fa = malheur_load();

    /* Extract prototypes */
    pr = proto_extract(fa, &as);

    /* Cluster prototypes and extrapolate */
    cluster_t *c = cluster_linkage(pr, 0);
    cluster_extrapolate(c, as);
    cluster_trim(c);

    /* Save prototypes */
    pn = cluster_get_prototypes(c, as, pr);
    if (save)
        farray_save_file(pn, mcfg.proto_file);
    farray_destroy(pn);

    /* Save rejected feature vectors */
    re = cluster_get_rejected(c, fa);
    if (save)
        farray_save_file(re, mcfg.reject_file);
    farray_destroy(re);

    /* Export clustering */
    export_cluster(c, pr, fa, as, output_file);
    
    /* Export shared n-grams */
    export_shared_ngrams(c, fa, output_file);

    /* Clean up */
    cluster_destroy(c);
    assign_destroy(as);
    farray_destroy(pr);
    farray_destroy(fa);
}

/**
 * Classify the given malware reports
 */
static void malheur_classify()
{
    assign_t *as;
    farray_t *fa, *pr, *re;

    /* Check for prototype file */
    if (access(mcfg.proto_file, R_OK))
        fatal("No prototype file for classifcation available");

    /* Load data */
    fa = malheur_load();

    /* Load prototypes */
    pr = farray_load_file(mcfg.proto_file);

    /* Apply classification */
    as = class_assign(fa, pr);

    /* Save rejected feature vectors */
    re = class_get_rejected(as, fa);
    if (save)
        farray_save_file(re, mcfg.reject_file);
    farray_destroy(re);

    /* Export classification */
    export_class(pr, fa, as, output_file);

    /* Clean up */
    assign_destroy(as);
    farray_destroy(pr);
    farray_destroy(fa);
}

/**
 * Classify the given malware reports
 */
static void malheur_increment()
{
    farray_t *pr = NULL, *tmp, *pn, *re;
    assign_t *as; 

    /* Load internal state */
    malheur_load_state();

    /* Load data including rejected stuff */
    farray_t *fa = malheur_load();
    if (!access(mcfg.reject_file, F_OK)) {
        tmp = farray_load_file(mcfg.reject_file);
        fa = farray_merge(fa, tmp);
    }

    /* Classification */
    if (!access(mcfg.proto_file, R_OK)) {
        pr = farray_load_file(mcfg.proto_file);

        /* Apply classification */
        as = class_assign(fa, pr);
        tmp = class_get_rejected(as, fa);
        
        /* Export results */
        export_increment1(pr, fa, as, output_file);
        
        /* Clean up */
        farray_destroy(fa);
        farray_destroy(pr);
        assign_destroy(as);
        fa = tmp;        
    } else {
        /* Export results */
        export_increment1(pr, fa, as, output_file);
    }

    /* Extract prototypes */
    pr = proto_extract(fa, &as);
    
    /* Cluster prototypes and extrapolate */
    cluster_t *c = cluster_linkage(pr, mstate.run + 1);
    cluster_extrapolate(c, as);
    cluster_trim(c);

    /* Save prototypes vectors */
    pn = cluster_get_prototypes(c, as, pr);
    if (save)
        farray_append_file(pn, mcfg.proto_file);

    /* Save rejeted feature vectors */
    re = cluster_get_rejected(c, fa);
    if (save)
        farray_save_file(re, mcfg.reject_file);

    /* Update state */
    mstate.run++;
    mstate.num_proto = pn->len;
    mstate.num_reject = re->len;

    /* Save state */
    if (save) 
        malheur_save_state();
    
    /* Export results */
    export_increment2(c, pr, fa, as, output_file);

    /* Clean up */
    cluster_destroy(c);
    assign_destroy(as);

    farray_destroy(re);
    farray_destroy(pn);
    farray_destroy(pr);
    farray_destroy(fa);
}

/**
 * Display information about internal state of Malheur
 */
static void malheur_info()
{
    /* Load internal state */
    if (!malheur_load_state()) {
        printf("No internal state stored in %s\n", malheur_dir);
        return;
    }

    printf("Internal state from %s\n", mcfg.state_file);
    printf("       Malheur run: %u\n", mstate.run);
    printf(" Stored prototypes: %u\n", mstate.num_proto);
    printf("  Rejected reports: %u\n", mstate.num_reject);

}

/**
 * Computes a distance matrix and saves the result to a file
 */
static void malheur_distance()
{
    /* Load data */
    farray_t *fa = malheur_load();

    /* Allocate distance matrix */
    double *d = malloc(fa->len * fa->len * sizeof(double));
    if (!d)
        fatal("Could not allocate similarity matrix");

    /* Compute distance matrix */
    farray_dist(fa, fa, d);

    /* Save distance matrix */
    export_dist(d, fa, output_file);

    /* Clean up */
    free(d);
    farray_destroy(fa);
}

/**
 * Computes a distance matrix for the prototypes and saves the result to a file
 */
static void malheur_protodist()
{
    farray_t *pr;

    /* Check for prototype file */
    if (access(mcfg.proto_file, R_OK)) 
        fatal("No prototype file for classifcation available");
    
    /* Load prototypes */
    pr = farray_load_file(mcfg.proto_file);
    if (verbose > 1)
        farray_print(pr);
    
    /* Allocate distance matrix */
    double *d = malloc(pr->len * pr->len * sizeof(double));
    if (!d)
        fatal("Could not allocate similarity matrix");

    /* Compute distance matrix */
    farray_dist(pr, pr, d);

    /* Save distance matrix */
    export_dist(d, pr, output_file);

    /* Clean up */
    free(d);
    farray_destroy(pr);
}

/**
 * Exits the malheur tool.
 */
static void malheur_exit()
{
    int lookup;

    /* Destroy feature lookup table */
    config_lookup_int(&cfg, "features.lookup_table", &lookup);
    if (lookup) {
        if (verbose > 0)
            ftable_print();
        ftable_destroy();
    }

    /* Destroy configuration */
    config_destroy(&cfg);
}

/**
 * Main function of Malheur
 * @param argc Number of arguments
 * @param argv Argument values
 * @return Exit code
 */
int main(int argc, char **argv)
{
    malheur_init(argc, argv);

    /* Perform action */
    switch (action) {
    case DISTANCE:
        malheur_distance();
        break;
    case PROTOTYPE:
        malheur_prototype();
        break;
    case CLUSTER:
        malheur_cluster();
        break;
    case CLASSIFY:
        malheur_classify();
        break;
    case INCREMENT:
        malheur_increment();
        break;
    case PROTODIST:
        malheur_protodist();
        break;
    case INFO:
        malheur_info();
    }

    malheur_exit();
    return EXIT_SUCCESS;
}
